torchgeo/datasets/geo.py

```diff
@@ -418,6 +420,61 @@ def __init__(
         self._crs = cast(CRS, crs)
         self._res = cast(float, res)
 
+    def list_files(self, filename_glob: Optional[str] = None) -> list[str]:
+        """Get list of files matching filename_glob.
+
+        Args:
+            filename_glob: Defaults to self.filename_glob
+        """
+        # self.paths can be single string (or pathlike) representing root directory.
+        # Make iterable
+        if isinstance(self.paths, str) or not isinstance(self.paths, Iterable):
+            paths = [self.paths]
+        else:
+            paths = cast(list[str], self.paths)
+
+        filename_glob = filename_glob or self.filename_glob
+
+        # using set to remove any duplicates if directories are overlapping
+        filepaths: set[str] = set()
+        for dir_or_file in paths:
+            if os.path.exists(dir_or_file):
+                if os.path.isfile(dir_or_file):
+                    filepaths.add(dir_or_file)
+                else:
+                    pathname = os.path.join(dir_or_file, "**", filename_glob)
+                    filepaths |= set(glob.iglob(pathname, recursive=True))
+            else:
+                filepaths |= self.handle_nonlocal_path(dir_or_file)
+
+        return list(filepaths)
+
+    def handle_nonlocal_path(self, path: str) -> set[str]:
```

Comment:
I don't want to do this just yet, can we save it for your other PR? I'm actually leaning towards using classmethods like `from_vsi` or `from_s3`, or just telling users to pass in a list of files if they want to use a remote server that can't be globbed. So for now, use something like:
```python
for path in paths:
    if os.path.isdir(path):
        # glob
    else:
        # add
```

Reply:
I'm ok with that. Now that `list_files` (or what it will be called) is implemented, it is much easier for users (me) to override this part of the code in order to support vsi.